### PR TITLE
[GSB] Only use matching superclass constraints in enumerated requirements

### PR DESCRIPTION
**Pull request by DougGregor**
Fixes SR-5165 / rdar://problem/32658705.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

File: lib/AST/GenericSignatureBuilder.cpp [hunk 1]
```diff
@@ -5092,14 +5095,21 @@ void GenericSignatureBuilder::enumerateRequirements(llvm::function_ref<
     // If we have a superclass, produce a superclass requirement
     if (equivClass->superclass && !equivClass->recursiveSuperclassType) {
       f(RequirementKind::Superclass, archetype, equivClass->superclass,
-        getBestConstraintSource<Type>(equivClass->superclassConstraints));
+        getBestConstraintSource<Type>(equivClass->superclassConstraints,
+                        [&](const Type &type) {
+                          return type->isEqual(equivClass->superclass) ||
+                            equivClass->superclass->hasError();
```
Comment (slavapestov) [inline lib/AST/GenericSignatureBuilder.cpp]: What's the significance of hasError() here?

**Comment by DougGregor** [comment]
Ugh, bad test case merge

File: lib/AST/GenericSignatureBuilder.cpp (hunk 1 above)
Comment (DougGregor) [inline lib/AST/GenericSignatureBuilder.cpp]: We may not have an exact match if there are errors. I've decided that I don't like this approach and will improve on this a bit.

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Pull request merged**